### PR TITLE
Port kernel vmmap to aglib

### PR DESCRIPTION
**Pull request by patryk4815**
Changes:
- moved `kernel_vmmap_via_monitor_info_mem`  to `aglib.kernel.vmmap`
- moved `kernel_vmmap_via_page_tables` to  `aglib.kernel.vmmap`
- fix weird response in lldb send_monitor
- Rewrite pt_gdb to aglib in `kernel_vmmap_via_page_tables`
- implement rv64 paging it allows to use `kernel_vmmap_via_page_tables`

### CODE REVIEW

File: pwndbg/aglib/vmmap.py [hunk 1]
```diff
@@ -20,9 +20,6 @@
 
 @pwndbg.lib.cache.cache_until("start", "stop")
 def get() -> Tuple[pwndbg.lib.memory.Page, ...]:
-    if pwndbg.dbg.is_gdblib_available():
```
Comment (patryk4815) [inline pwndbg/aglib/vmmap.py:23]: redundant, this is done inside `pwndbg.dbg.selected_inferior().vmmap()`

File: pwndbg/aglib/kernel/vmmap.py [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+        return int(getattr(pwndbg.aglib.regs, register_name))
+
+
+@pwndbg.lib.cache.cache_until("stop")
+def kernel_vmmap_via_page_tables() -> Tuple[pwndbg.lib.memory.Page, ...]:
+    if not pwndbg.aglib.qemu.is_qemu_kernel():
+        return ()
+
+    # If paging is not enabled, we shouldn't attempt to parse page tables
+    if not pwndbg.aglib.kernel.paging_enabled():
```
Comment (patryk4815) [inline pwndbg/aglib/kernel/vmmap.py]: this raise exception

File: pwndbg/aglib/kernel/__init__.py [hunk 3]
```diff
@@ -649,6 +650,10 @@ def paging_enabled() -> bool:
         return x86_64Ops.paging_enabled()
```
Comment (disconnect3d) [inline pwndbg/aglib/kernel/__init__.py:650]: Random note: we should reorder this to x86-64, aarch64, rv64, i386

File: pwndbg/aglib/kernel/__init__.py [hunk 4]
```diff
@@ -649,6 +650,10 @@ def paging_enabled() -> bool:
...
     elif arch_name == "aarch64":
         return Aarch64Ops.paging_enabled()
+    elif arch_name == "rv64":
+        # https://starfivetech.com/uploads/u74_core_complex_manual_21G1.pdf
+        # page 41, satp.MODE, bits: 60,61,62,63
+        return int(pwndbg.aglib.regs.satp) & (BIT(60) | BIT(61) | BIT(62) | BIT(63)) != 0
```
Comment (disconnect3d) [inline pwndbg/aglib/kernel/__init__.py:657]: image

Comment (disconnect3d) [inline pwndbg/aglib/kernel/__init__.py:657]:
```suggestion
        # "When satp.MODE=0x0, supervisor virtual addresses are equal to supervisor physical addresses"
        return int(pwndbg.aglib.regs.satp) & (BIT(60) | BIT(61) | BIT(62) | BIT(63)) != 0
```

File: pwndbg/aglib/kernel/vmmap.py [hunk 5]
```diff
@@ -0,0 +1,257 @@
...
+import pwndbg.aglib.arch
+import pwndbg.aglib.kernel
+import pwndbg.aglib.qemu
+import pwndbg.aglib.regs
+import pwndbg.color.message as M
+import pwndbg.lib.cache
+import pwndbg.lib.memory
+
+
+# Most of QemuMachine code was copied from:
```
Comment (disconnect3d) [inline pwndbg/aglib/kernel/vmmap.py]:
```suggestion
# Most of QemuMachine code was inherited from gdb-pt-dump thanks to Martin Radev (@martinradev)
# on the MIT license, see:
```

File: pwndbg/aglib/kernel/vmmap.py [hunk 6]
```diff
@@ -0,0 +1,258 @@
...
+# Most of QemuMachine code was inherited from gdb-pt-dump thanks to Martin Radev (@martinradev)
+# on the MIT license, see:
+# https://github.com/martinradev/gdb-pt-dump/blob/21158ac3f9b36d0e5e0c86193e0ef018fc628e74/pt_gdb/pt_gdb.py#L11-L80
+class QemuMachine(Machine):
+    def __init__(self):
+        super().__init__()
+        self.pid = QemuMachine.get_qemu_pid()
+        self.file = None
+        self.file = os.open(f"/proc/{self.pid}/mem", os.O_RDONLY)
+        self.mem_size = os.fstat(self.file).st_size
```
Comment (disconnect3d) [inline pwndbg/aglib/kernel/vmmap.py]: Does this make sense?
```sh
(.venv) root@pwndbg:~# ls -la /proc/self/mem
-rw------- 1 root root 0 Dec 18 13:00 /proc/self/mem
(.venv) root@pwndbg:~# stat /proc/self/mem
  File: /proc/self/mem
  Size: 0         	Blocks: 0          IO Block: 1024   regular empty file
Device: 0,22	Inode: 29332658    Links: 1
Access: (0600/-rw-------)  Uid: (    0/    root)   Gid: (    0/    root)
Access: 2024-12-18 13:00:15.561108327 +0000
Modify: 2024-12-18 13:00:15.561108327 +0000
Change: 2024-12-18 13:00:15.561108327 +0000
 Birth: -
```

Comment (disconnect3d) [reply pwndbg/aglib/kernel/vmmap.py]:
```py
In [4]: f=os.open('/proc/self/mem', os.O_RDONLY)

In [5]: os.fstat(f).st_size
Out[5]: 0
```

cc: @martinradev

Comment (disconnect3d) [reply pwndbg/aglib/kernel/vmmap.py]: https://github.com/martinradev/gdb-pt-dump/pull/37

File: pwndbg/aglib/kernel/vmmap.py [hunk 7]
```diff
@@ -0,0 +1,258 @@
...
+        # We add a chardev file backend (we dont add a fronted, so it doesn't affect
+        # the guest). We can then look through proc to find which process has the file
+        # open. This approach is agnostic to namespaces (pid, network and mount).
+        chardev_id = "gdb-pt-dump" + "-" + "".join(random.choices(string.ascii_letters, k=16))
+        with tempfile.NamedTemporaryFile() as tmpf:
+            pwndbg.dbg.selected_inferior().send_monitor(
+                f"chardev-add file,id={chardev_id},path={tmpf.name}"
+            )
+            pid_found = QemuMachine.search_pids_for_file(pids, tmpf.name)
+            pwndbg.dbg.selected_inferior().send_monitor(f"chardev-remove {chardev_id}")
```
Comment (disconnect3d) [inline pwndbg/aglib/kernel/vmmap.py:75]: Lol this is a smart hack to detect the given QEMU process PID.

Maybe we should implement a monitor command to just return QEMU pid? :D

File: pwndbg/aglib/kernel/vmmap.py [hunk 8]
```diff
@@ -0,0 +1,258 @@
...
+                )
+            )
+        return ()
+
+    lines = monitor_info_mem.splitlines()
+
+    # Handle disabled PG
+    # This will prevent a crash on abstract architectures
+    if len(lines) == 1 and lines[0] == "PG disabled":
+        return ()
```
Comment (disconnect3d) [inline pwndbg/aglib/kernel/vmmap.py:219]: What's PG disabled, is it paging? Is it only on one or all archs?

Any chance we can support physical addresses here too? or should we handle it elsewhere?

File: pyproject.toml [hunk 9]
```diff
@@ -116,7 +116,7 @@ module = [
     "r2pipe",
     "rzpipe",
     "rich.*",
-    "pt_gdb",
+    "pt.*",
```
Comment (disconnect3d) [inline pyproject.toml:119]: still needed or not?

**Review by disconnect3d — APPROVED** [review]

File: pwndbg/aglib/kernel/vmmap.py (hunk 6 above)
Comment (disconnect3d) [inline pwndbg/aglib/kernel/vmmap.py]:
```suggestion
```

**Comment by martinradev** [comment]
Is there an alternative to copy-pasting the gdb-pt-dump code, and potentially something I could assist with?

If the code becomes duplicated, code maintainability for both projects will suffer:
- fixes and new functionality will not be making it into this downstream, unless one of you decides to copy-paste the code again to freshen it up.
- potential fixes or new functionality added to this downstream copy may be forgotten to be merged back to the upstream.
- any error messages, related to the downstream copy of gdb-pt-dump, may be misattributed to the upstream gdb-pt-dump project that will then require me to fix or only copy-paste into the downstream.

What would be required from gdb-pt-dump side to have the code live in a single place, but have things bundled for a release?

**Pull request merged**